app/services/hyrax/analytics/google_analytics.rb

```diff
@@ -1,76 +1,114 @@
+        )
+      end
+      private_class_method :report_request
 
-        def initialize(config)
-          @config = config
+      def self.format_date(date)
+        if date.is_a?(String)
+          date
+        else
+          date.strftime('%Y-%m-%d')
         end
+      end
+      private_class_method :format_date
 
-        # @return [Boolean] are all the required values present?
-        def valid?
-          config_keys = @config.keys
-          REQUIRED_KEYS.all? { |required| config_keys.include?(required) }
+      def self.stats_rows(response, query_params)
+        output = []
+        response.reports.first.data.rows.each do |row|
+          output_row = OpenStruct.new
+          output << openstruct_row(row.to_h, output_row, query_params)
         end
+        output
+      end
+      private_class_method :stats_rows
 
-        REQUIRED_KEYS.each do |key|
-          class_eval %{ def #{key};  @config.fetch('#{key}'); end }
+      def self.openstruct_row(row, output_row, query_params)
+        Rails.logger.error 'Hyrax::Analytics.GoogleAnalytics - Bad Google Analytics data' if row[:dimensions].blank? || row[:metrics].blank?
+        query_params[:dimensions].zip(row[:dimensions]).each do |term, value|
+          output_row[term] = value
         end
+        query_params[:metrics].zip(row[:metrics].first[:values]).each do |term, value|
+          output_row[term] = value
+        end
+        output_row
+      end
+      private_class_method :openstruct_row
 
-        # Return a Google Analytics Reporting Service
-        def self.profile
-          return unless config.valid?
-          unless File.exist?(config.privkey_path)
-            raise "Private key file for Google Analytics was expected at '#{config.privkey_path}', but no file was found."
-          end
-          analytics = Google::Apis::AnalyticsreportingV4::AnalyticsReportingService.new
-          credentials = Google::Auth::ServiceAccountCredentials.make_creds(json_key_io: File.open(config.privkey_path))
-          credentials.scope = 'https://www.googleapis.com/auth/analytics.readonly'
-          analytics.authorization = credentials.fetch_access_token!({})["access_token"]
-          [analytics, 'ga:' + config.view_id.to_s]
+      # Return an authorized Google Analytics Reporting Service
+      def self.setup_and_authorize
+        unless File.exist?(config['privkey_path'])
+          raise "Private key file for Google Analytics was expected at '#{config['privkey_path']}', but no file was found."
         end
+        analytics = Google::Apis::AnalyticsreportingV4::AnalyticsReportingService.new
+        credentials = Google::Auth::ServiceAccountCredentials.make_creds(json_key_io: File.open(config['privkey_path']))
+        credentials.scope = 'https://www.googleapis.com/auth/analytics.readonly'
+        analytics.authorization = credentials.fetch_access_token!({})["access_token"]
+        analytics
```

Comment:
Saw your Slack message about timeout concerns re: memoization. Which does make perfect sense. But I wonder if the auth library has any features that can help mitigate that problem? I see it supports token stores (file and redis) but I'm not sure, after a cursory glance, where there's an easy way to ask if the current credentials/token are still valid and then refresh if necessary.

Reply:
Will spin a new ticket to figure this out. It's not easy to see if the gem takes care of refresh tokens for you or not.